Hash source text inside DebuggerSource

kannanvijayan opened the pull request.

kannanvijayan committed:
Hash source text inside DebuggerSource.

kannanvijayan commented:
Draft patch.  Something's not quite right and I'm not seeing printf output (but page loads fast!).

Currently it just generates a unique string with an incrementing counter as the "hash".  If we verify that this saves a lot of speed we can put a real hash function inside.  We'd probably want to design the hashing semantics around hashing char16s directly (should be fine as long as it's consistently applied).  That'll allow for a cheaper conversion from the sources, which may possibly be in Latin1 (8-bit).  These can be simply promoted to char16s directly.

We can have a discussion about that after I get the patch verified.

gideonred commented:
Great. I still wonder if the bottle neck is the calculation versus the reading of the file. If it is the reading of the file then no amount of hash algorithm will help.

Here's some observations I made quickly looking at this code:
- We calculate three different layers of hashing for the same file to deal with hash collisions. Do we have to calculate all three upfront, or can we delay it to when there is actually a collision?
- The browser calculates the hash, but never uses the hash. It's sent to the backend. We could offload the calculation from the already struggling browser to the backend and scale up the backend.

kannanvijayan commented:
Reading the contents shouldn't be an issue since the source text has just been parsed and should be in memory anyway.  I got this working, but the aggregate perf improvements are nice, but still leaves a bunch on the table.

Average time-to-first-console goes down to 5s with a dummy-hash.  Note that I'm not actually hashing contents at all right now.  So the final version of this patch would be slower than this.

What the timings suggest is that URL-hashing is also a cost, for some reason?  It really shouldn't be.

I'm going to measure the cost of hashing the URLs directly again.

```
With dummy-hash patch, no other changes
5.253
5.051
4.914
5.406

Return before send-async-message.
  - With patch, but modified to return before the `sendAsyncMessage` call in `setSourceMap`
3.641
3.988
3.753
3.591

Move hashing of url contents before return
  - With patch, returning before `sendAsyncMessage`, but computing the API-hashes for the URLs before the return.
4.74
4.94
4.967
4.826

Send async, but with no hashing
  - With patch, performing the `sendAsyncMessage`, but bypassing the API-hashes of the URLs.
4.208
4.388
4.362
3.789
4.038
4.171
```

loganfsmyth commented on devtools/server/actors/replay/module.js:
```diff
@@ -265,7 +265,7 @@ function setSourceMap({
 
     // Attempt to be more specific by matching on the script's URL and content.
-    targetContentHash: typeof objectText === "string" ? makeAPIHash(objectText) : undefined,
+    targetContentHash: objectHash,
```
`makeAPIHash` prepends a `sha256:` that appears to be missing now FYI

kannanvijayan replied:
So the final patch integrates xxhash instead (it's apparently faster).  Should we use 'xxh128:' or something?

loganfsmyth replied:
I think I initially missed that we were still considering changing the hash to something other than SHA256. We'll need backend support for the new type of hash, which isn't a trivial amount of work. I'm very tempted to say that that might be only slightly less work than supporting `sourceId` in the API instead. The more of think about this, the less I think adding a new type of hash is a worth the effort relative to the other option.

kannanvijayan replied:
Ok I've switched it back to SHA256 on canonicalised UTF8 text as before.  I'll double-check the hashes are the same as it would have been in prior code, but we should be good to go on this front.

loganfsmyth commented on js/src/debugger/Source.cpp:
```diff
@@ -202,6 +203,42 @@ bool DebuggerSource::CallData::ToNative(JSContext* cx, unsigned argc,
+      return nullptr;
+    }
+    if (!hasSourceText) {
+      return NewStringCopyZ<CanGC>(cx_, "[no source]");
```
Seems like we should return `nullptr` rather than something that isn't a hash.

kannanvijayan replied:
nullptr is hard to distinguish from a failure-to-allocate.  I could return the empty string I guess?  Or wrap the return value in a `js::Maybe`

loganfsmyth replied:
Returning empty string and having the JS logic skip the call to upload the sourcemap is fine then.

kannanvijayan replied:
Haven't done it yet, but will do.

loganfsmyth commented on js/src/debugger/Source.cpp:
```diff
@@ -202,6 +203,42 @@ bool DebuggerSource::CallData::ToNative(JSContext* cx, unsigned argc,
+  }
+
+  ReturnType match(Handle<WasmInstanceObject*> instanceObj) {
+    return NewStringCopyZ<CanGC>(cx_, "[wasm]");
```
Same here.

kannanvijayan commented:
Couple minor issues to address (e.g. "xxh128:" prefix, removing commented-out debug-print lines).

But otherwise this is ready to talk about at least.  I'd encourage people to try it out on their own machine, but on the `OnConsoleAPI` measurement, this gets us down to 4s from 7s (start of page load to first console log from the page).

This patch, in addition to doing source-hashing inside the ScriptSource object, refuses to hash "data:" source-map URLs and sends undefineds for those instead.

I need to run an errand at the moment but I'll do some direct measurements on the xxhash impl itself (the `source.hash` call runs it).

In the meantime, @loganfsmyth @bhackett1024 things to discuss:

1. Are we cool with xxhash?
2. What about that BSD license?
3. Why is it ok to send "undefined" for the source map url hash for "data:" urls?

kannanvijayan commented:
Jason suggested putting SHA256 back into the source hashing in this modified approach.

I'll work on getting that implemented today.  It's kind of awkward to use gecko stuff (nsICryptoHash in particular) from spidermonkey, so I might just make a new file with the SHA256 implementation.

The problem with preserving our current hashing behaviour is that it forces me to to introduce conversions.  Let me explain in more detail:

Source data is kept internally in a `ScriptSource` object, which may have received the source data either from network (in UTF8 format) or from an internal JS string (e.g. with "new Function()" or "eval").  It keeps this source data in its original format internally - holding a pointer to either a UTF8 char* or a JS char16*.

The semantics of our current hashing behaviour is that all sources UTF8 encodings are hashed at the byte level.  To do this properly, I'd have to handle the char16-source case specially: convert it to UTF8 first, and then pass it to the hash function.

Yesterday I measured the number of "big" sources with char16 sources showing up in our test case, and there are many of them.  I was hoping that all the "big" sources would end up having UTF8 text, but it wasn't the case.  There were several >200k sources that were kept internally as char16 strings.

So I'll go ahead and implement a "proper" patch with SHA256 that operates on UTF8 source only, and measure the performance we lose for a) the return back to SHA256, and b) converting char16 sources to UTF8 before hashing them

kannanvijayan commented:
Updated patch.  Added a simple SHA256 implementation and use it within the source hashing.  This time I am making sure to convert the char16_t source strings into UTF8 before hashing.  All hashes should be consistent with our prior approach but I haven't verified it yet.  Will be doing that soon.

In the meantime, switching to SHA256 and canonicalizing the source to UTF8 pushes the OnConsoleAPI metric up to about 4.3 seconds.  Visually, it _feels_ reasonably faster than the slow case, but also noticeably slower than the case where we skip the entire `setSourceMap` logic and hashing.

kannanvijayan commented:
Just speculation, but I think part of the high performance cost of the hashing previously was the fact that we were allocating both a JSString and a ByteArray in javascript managed memory.  Not only were we paying the conversion costs for hundred-kbs to megabytes of text, every "large" source would stress the JS heap with two large allocations that would go straight to tenured memory.

This might well have been having cascading consequences for performance down the line if the browser was induced to perform more major GCs because of this behaviour.

jasonLaster commented on devtools/server/actors/replay/module.js:
```diff
@@ -802,6 +803,7 @@ function consoleAPIMessageLevel({ level }) {
 }
 
 function OnConsoleAPICall(message) {
+  // dump(`KVKV OnConsoleApi ${(Date.now() - startTime) / 1000}\n`);
```
we can remove this

jasonLaster reviewed (commented):
Looks good to me

bhackett1024 commented on js/src/vm/JSScript.cpp:
```diff
@@ -2184,6 +2186,61 @@ JSLinearString* ScriptSource::substring(JSContext* cx, size_t start,
+  return NewStringCopyUTF8N<CanGC>(cx, JS::UTF8Chars(buf, 64));
+}
+
+static int KvKvCounter = 0;
```
rm

bhackett1024 commented on js/src/vm/SHA256.cpp:
```diff
@@ -0,0 +1,180 @@
+*/
+
+// Copied from
+// https://github.com/System-Glitch/SHA256/blob/388853aaa3e1d91d37d3bfa5abadc0a4a70283fa/include/SHA256.h
```
this looks like the wrong file

kannanvijayan commented:
Latest push addresses final review comments, all nits.  The prior patch was passing in build-test dashboard:

```
Gecko Build/Test Branch kannan/run-382-hash-source-text-within-debuggersource bd99a4750a16 2022-06-26T01:12:32.470Z

Build Gecko macOS: passed [log](https://github.com/RecordReplay/build-test-dashboard/blob/master/4a11a800-27ca-435d-b26a-1c7aa578b856)

Run Tests macOS: passed [log](https://github.com/RecordReplay/build-test-dashboard/blob/master/6fe177b5-34ad-488b-bdfb-60e83b220356)

Build Gecko linux: passed [log](https://github.com/RecordReplay/build-test-dashboard/blob/master/71d7a814-d5ce-462a-b49b-0fffcc61c29b)

Run Tests linux: passed [log](https://github.com/RecordReplay/build-test-dashboard/blob/master/68f21e53-36aa-4faf-9ef6-573749370413)

Build Gecko windows: passed [log](https://github.com/RecordReplay/build-test-dashboard/blob/master/5cf093e5-a8e5-4487-92d0-141f8ebe9f37)

Run Tests windows: passed [log](https://github.com/RecordReplay/build-test-dashboard/blob/master/0fd3b30d-11a0-4049-ad60-69fe90c14a54)
```

I think this is good enough to land.
